### PR TITLE
feat: support custom resizes for StatefulSet

### LINKED ISSUE #13641: Support custom StatefulSet resizes in webhook
**What would you like to be added**:
Add support for custom StatefulSet resizing (e.g., scaling replicas from 3 to 5) while preserving job queueing semantics, resolving the TODO in `pkg/controller/jobs/statefulset/statefulset_webhook.go`.

**Why is this needed**:
Currently, Kueue's StatefulSet webhook only allows scaling down to zero (suspending) or scaling up from zero (resuming). Distributed inference or data processing workloads often use StatefulSets and need the ability to dynamically resize (scale out/in) without being fully suspended and requeued from zero. (This touches on the gap identified in issue #3279).

**Completion requirements**:

This enhancement requires the following artifacts:

- [ ] Design doc
- [ ] API change
- [ ] Docs update

The artifacts should be linked in subsequent comments.

### PR DESCRIPTION
**Pull request by Vaishnav88sk**
#### What type of PR is this?
/kind feature 
/area integrations

#### What this PR does / why we need it:
This PR enables dynamic scaling (custom resizes) for StatefulSet workloads. 

Changes include:
- **Webhook Updates:** Removes the blanket restriction on all non-zero to non-zero `replicas` changes. It now allows custom resizes, but actively blocks scale-out operations (increasing replicas) for *already admitted* workloads, since Kueue currently does not support dynamic quota increases for admitted jobs.
- **Reconciler Updates:** Syncs replica changes back to the Workload:
  - For **pending workloads**, it directly updates `wl.Spec.PodSets[0].Count` so the workload retains its queue position with the updated quota request.
  - For **admitted workloads (scale-in)**, it computes the unused quota and triggers `UpdateReclaimablePods` to release it back to the queue for other workloads to use.

#### Which issue(s) this PR fixes:

Fixes #13641 

#### Does this PR introduce a user-facing change?

```release-note
Add support for dynamic custom resizing for StatefulSet workloads (scale-in and scale-out).
```

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved StatefulSet scaling reconciliation for admitted workloads.
  * Scale-downs now track reclaimed pods while preserving admitted workload replica counts.
  * Pending workloads can scale up and update their requested pod counts.
  * Scale-outs beyond admitted capacity are blocked or safely reverted.
  * Improved validation for scaling from zero while pods terminate or a prior workload remains active.

* **Tests**
  * Expanded coverage for pending and admitted scaling, rollback behavior, reclaimed pods, and webhook validation.

### CODE REVIEW

File: pkg/controller/jobs/statefulset/statefulset_reconciler.go [hunk 1]
```diff
@@ -260,6 +289,12 @@ func (r *Reconciler) reconcileWorkload(ctx context.Context, sts *appsv1.Stateful
 		}
 	}
 
+	if shouldUpdateReclaimablePods {
+		if err := workload.UpdateReclaimablePods(ctx, r.client, wl, reclaimablePods); err != nil {
+			return err
+		}
+	}
```
Comment (mimowo) [inline pkg/controller/jobs/statefulset/statefulset_reconciler.go]: Can you move it to a helper which reflects the intention of the code by name and comment. Currently it seems pretty involing to understand without some commentary

**Review by mimowo — COMMENTED** [review]
/hold

I think to claim the support for custom resizes we need e2e tests, and maybe some integration tests. I remember we iterated with @mbobrovskyi a couple of times and concluded this is rather hard problem. Sure, I'm happy to support arbitrary scaling, but without higher-level tests we cannot claim this really works.

**Comment by mimowo** [comment]
cc @mbobrovskyi ptal

File: pkg/controller/jobs/statefulset/statefulset_reconciler.go (hunk 1 above)
Comment (Vaishnav88sk) [reply pkg/controller/jobs/statefulset/statefulset_reconciler.go]: Makes sense

**Comment by Vaishnav88sk** [comment]
> I think to claim the support for custom resizes we need e2e tests, and maybe some integration tests. I remember we iterated with @mbobrovskyi a couple of times and concluded this is rather hard problem. Sure, I'm happy to support arbitrary scaling, but without higher-level tests we cannot claim this really works.

Given the edge cases around admission webhooks and controller race conditions during scaling, proving this via tests is important.

I will add comprehensive integration tests covering:
1. Scale-up being properly rejected/reverted if the quota hasn't explicitly admitted the new size (TOCTOU mitigation).
2. Scale-down properly creating `ReclaimablePods` to free up quota.
3. Pre-admission scaling seamlessly updating the Workload's `PodSet` count.

Since you mentioned iterating on this with @mbobrovskyi and finding it difficult in the past, are there any *specific* race conditions or edge cases you discovered back then that I should explicitly cover in the e2e/integration tests? I want to make sure I address your exact concerns.

File: pkg/controller/jobs/statefulset/statefulset_webhook.go [hunk 2]
```diff
@@ -185,33 +185,32 @@ func (wh *Webhook) ValidateUpdate(ctx context.Context, oldSTSObj, newSTSObj *app
...
-				allErrs = append(allErrs, field.Forbidden(replicasPath, "scaling down is still in progress"))
-			} else {
-				// Block if workload is still being deleted (exists without OnHold).
-				// If the workload is on hold, it is intentionally kept alive during
-				// scale-to-zero and will be re-admitted on scale-up.
-				wlName, err := findWorkloadName(ctx, wh.client, oldSTSObj)
-				if err != nil {
-					return nil, err
+			if newReplicas != 0 && oldReplicas != 0 {
+				if wlExists && workload.IsAdmitted(&wl) && newReplicas > oldReplicas {
```
Comment (mbobrovskyi) [inline pkg/controller/jobs/statefulset/statefulset_webhook.go]:
```suggestion
				if wlExists && workload.HasQuotaReservation(&wl) && newReplicas > oldReplicas {
```
Do we need to check for QuotaReservation?

File: pkg/controller/jobs/statefulset/statefulset_reconciler.go [hunk 3]
```diff
@@ -249,6 +249,35 @@ func (r *Reconciler) reconcileWorkload(ctx context.Context, sts *appsv1.Stateful
...
+					Name:  wl.Spec.PodSets[0].Name,
+					Count: wl.Spec.PodSets[0].Count - replicas,
+				},
+			}
+		case replicas > wl.Spec.PodSets[0].Count:
+			// TOCTOU mitigation: If the workload was admitted concurrently with a scale-out,
+			// the webhook may have allowed it because the Workload was pending at the time of check.
+			// Revert the StatefulSet replicas to match the admitted workload's quota.
+			count := wl.Spec.PodSets[0].Count
+			sts.Spec.Replicas = &count
```
Comment (mbobrovskyi) [inline pkg/controller/jobs/statefulset/statefulset_reconciler.go]:
```suggestion
			sts.Spec.Replicas = new(wl.Spec.PodSets[0].Count)
```

File: pkg/controller/jobs/statefulset/statefulset_reconciler.go [hunk 4]
```diff
@@ -260,6 +289,12 @@ func (r *Reconciler) reconcileWorkload(ctx context.Context, sts *appsv1.Stateful
 		}
 	}
 
+	if shouldUpdateReclaimablePods {
+		if err := workload.UpdateReclaimablePods(ctx, r.client, wl, reclaimablePods); err != nil {
+			return err
+		}
```
Comment (mbobrovskyi) [inline pkg/controller/jobs/statefulset/statefulset_reconciler.go]: I think we can move it to switch.

Comment (Vaishnav88sk) [reply pkg/controller/jobs/statefulset/statefulset_reconciler.go]: I've moved this logic into the extracted `syncReplicas` helper method as an early-return check. This flattens the logic and avoids unnecessarily nesting the `switch` statement.

File: pkg/controller/jobs/statefulset/statefulset_reconciler.go (hunk 3 above)
Comment (Vaishnav88sk) [reply pkg/controller/jobs/statefulset/statefulset_reconciler.go]: Done

File: pkg/controller/jobs/statefulset/statefulset_webhook.go (hunk 2 above)
Comment (Vaishnav88sk) [reply pkg/controller/jobs/statefulset/statefulset_webhook.go]: Updated

File: pkg/controller/jobs/statefulset/statefulset_reconciler.go [hunk 5]
```diff
@@ -249,6 +249,35 @@ func (r *Reconciler) reconcileWorkload(ctx context.Context, sts *appsv1.Stateful
 		shouldUpdate = true
 	}
 
+	var shouldUpdateReclaimablePods bool
+	var reclaimablePods []kueue.ReclaimablePod
+	if replicas > 0 && len(wl.Spec.PodSets) == 1 && wl.Spec.PodSets[0].Count != replicas {
+		switch {
+		case !workload.IsAdmitted(wl):
```
Comment (mbobrovskyi) [inline pkg/controller/jobs/statefulset/statefulset_reconciler.go]:
```suggestion
		case !workload.HasQuotaReservation(wl):
```
And here, right?

Comment (Vaishnav88sk) [reply pkg/controller/jobs/statefulset/statefulset_reconciler.go]: Yeah

File: pkg/controller/jobs/statefulset/statefulset_reconciler.go (hunk 1 above)
Comment (Vaishnav88sk) [reply pkg/controller/jobs/statefulset/statefulset_reconciler.go]: Done — extracted the whole replica-sync logic into a `syncReplicas` helper with comments explaining each case (pre-admission update, scale-down reclaim, TOCTOU revert).

**Comment by mimowo** [comment]
> Since you mentioned iterating on this with @mbobrovskyi and finding it difficult in the past, are there any specific race conditions or edge cases you discovered back then that I should explicitly cover in the e2e/integration tests? I want to make sure I address your exact concerns.

The main problematic I remember was scale down followed immediately by scale up, and vice versa, before the system has time to settle (as the operations span multiple objects).

Maybe @mbobrovskyi can share more stories

File: pkg/controller/jobs/statefulset/statefulset_reconciler.go [hunk 6]
```diff
@@ -485,3 +494,52 @@ func (h *podHandler) handle(obj client.Object, q workqueue.TypedRateLimitingInte
...
+		sts.Spec.Replicas = new(wl.Spec.PodSets[0].Count)
+		if err := r.client.Update(ctx, sts); err != nil {
+			return false, false, err
+		}
+		// The StatefulSet was updated, which will trigger a new reconciliation.
+		return false, true, nil
+	}
+
+	return false, false, nil
+}
```
[the remark this answers is not in the document]
Comment (Vaishnav88sk) [reply pkg/controller/jobs/statefulset/statefulset_reconciler.go:553]: done

**Comment by Vaishnav88sk** [comment]
/retest
infra

**Comment by thc1006** [comment]
Heads-up on a cross-PR interaction rather than a review of the change itself, since it only shows up once both land.

`Reconcile` runs the Pod work and `reconcileWorkload` as two goroutines on one `errgroup`, and both are handed the same `*appsv1.StatefulSet`. That is fine today because `reconcileWorkload` only reads it. This PR adds

```go
wlUpdated, stopReconcile, err := r.syncReplicas(ctx, sts, wl, replicas)
```

inside `reconcileWorkload`, and `syncReplicas` writes the field:

```go
sts.Spec.Replicas = new(wl.Spec.PodSets[0].Count)
if err := r.client.Update(ctx, sts); err != nil {
```

`syncQueueLabel`, on the other goroutine, opens with

```go
if sts == nil || ptr.Deref(sts.Spec.Replicas, 1) == 0 {
```

so it reads exactly what this writes. Same field, two goroutines, no synchronisation. `-race` would not see it on either branch alone, only on the merged tree.

The cheap fix is a deep copy for the Workload branch before the goroutines start, so each side owns its own StatefulSet. I have a PR in flight (#14117) that moves the queue-label sync onto the Pod branch, which is what puts the read there, so I am happy to add the copy on my side instead if that lands first. Either way it seemed worth writing down now rather than finding it after both merge.

This comment was written in part with the assistance of generative AI. Bcuz my English might be sucks.

**Comment by tenzen-y** [comment]
Hi @Vaishnav88sk can you rebase this PR onto the main branch?

**Comment by Vaishnav88sk** [comment]
> Hi @Vaishnav88sk can you rebase this PR onto the main branch?

Sorry team. Had some busy days. Now active again. 👍🏻
